### PR TITLE
Add Exp and Log RasterBlocks 

### PR DESCRIPTION
**Pull request by arjanverkerk**

### CODE REVIEW

File: dask_geomodeling/raster/elemwise.py [hunk 1]
```diff
@@ -746,3 +749,81 @@ def process(process_kwargs, *args):
...
 
         return {"values": values, "no_data_value": fillvalue}
+
+
+class BaseLogExp(BaseSingle):
+    """
+    Base Block for Exp, Log and Log10 Blocks.
+    """
+    def __init__(self, x):
+        if x.dtype == np.dtype("bool"):
```
Comment (caspervdw) [inline dask_geomodeling/raster/elemwise.py:759]: I think it is good to warn the users if the input is boolean dtype. It makes no sense to take Log or Exp from that.

Still, numpy just interprets it as 0 and 1 and nicely gives you back some numbers. What were your deliberations on this check?

File: dask_geomodeling/raster/elemwise.py [hunk 2]
```diff
@@ -746,3 +749,81 @@ def process(process_kwargs, *args):
...
+        super(BaseLogExp, self).__init__(x)
+
+    def get_sources_and_requests(self, **request):
+        process_kwargs = {"dtype": self.dtype.name, "fillvalue": self.fillvalue}
+        return [(process_kwargs, None), (self.args[0], request)]
+
+    @property
+    def dtype(self):
+        # use at least float32
+        return np.result_type(np.float32, *self.args)
```
Comment (caspervdw) [inline dask_geomodeling/raster/elemwise.py:770]: Numpy uses at least np.float16

```
In [6]: np.log10(np.array([1], dtype=np.float16))                             
Out[6]: array([0.], dtype=float16)

In [7]: np.log10(np.array([1], dtype=np.float32))                             
Out[7]: array([0.], dtype=float32)

In [9]: np.log10(np.array([1], dtype=np.int8))                                
Out[9]: array([0.], dtype=float16)

In [10]: np.log10(np.array([1], dtype=np.int16))                              
Out[10]: array([0.], dtype=float32)
```

But that is a bit weird dtype. I am :+1: on changing this to float32 in dask-geomodeling.

File: dask_geomodeling/raster/elemwise.py [hunk 3]
```diff
@@ -746,3 +749,81 @@ def process(process_kwargs, *args):
...
+
+    @property
+    def fillvalue(self):
+        return get_dtype_max(self.dtype)
+
+
+class Exp(BaseLogExp):
+    """
+    Return e raised to the power of the raster values.
```
Comment (caspervdw) [inline dask_geomodeling/raster/elemwise.py:779]: Good to note that out-of-range values (like Exp(1000000)) will become nodata as well

**Review by caspervdw — APPROVED** [review]
Some questions and minor docstring suggestion, for the rest it looks great!

Can be merged once CI agrees

File: dask_geomodeling/raster/elemwise.py (hunk 1 above)
Comment (arjanverkerk) [reply dask_geomodeling/raster/elemwise.py:759]: I know numpy doesn't complain, but dask_geomodeling complaining might stop someone from doing something unintended.

File: dask_geomodeling/raster/elemwise.py (hunk 3 above)
Comment (arjanverkerk) [reply dask_geomodeling/raster/elemwise.py:779]: Except yet it doesn't ;-) - do we have this kind of protections for other blocks as well? Coming to think of it, what do you think of putting this non-finite masking in the `wrap_math_process_func`? It took my a while to find out that the MemorySource and the MockRaster already do this.

And then, do you prefer out-of-range values on individual block docstrings, or shall we make it a general not, and where should it go then?

Comment (caspervdw) [reply dask_geomodeling/raster/elemwise.py:782]: I like to lift this `isfinite` masking to the `wrap_math_process_func`. Also for Multiply the issue will be possible definitely

**Comment by caspervdw** [comment]
Looks great! I really like the 1-line implementation of the blocks :)

**Pull request merged**